### PR TITLE
Fix/kendra webcrawler fixes

### PR DESCRIPTION
**Pull request by rajc-aws**
*Description of changes:*

- fixed Kendra Webcrawler data source sync issue by adding support to check for Kendra webcrawler data source {status} before initiating sync
- fixed issue where the Kendra Webcrawler data source {settings} page in Kendra console was resulting in a blank page - by updating configuraton parameters for data source create and update steps
- updated API Gateway Integration Request setting (for Kendra Webcrawler POST method) to include HTTP header to support asynchronous execution of the backend Lambda function
- updated status reporting for Kendra Webcrawler
- updated build setup in makefiles
- updated changelog

Submitting this PR with these changes for review.

### CODE REVIEW

File: lambda/kendra-webcrawler-schedule-updater/Makefile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 NAME=$(shell basename $(shell pwd))
 DST=../../build/lambda/$(NAME).zip
-RESOURCES := $(shell  find . | grep -v node_modules | grep -v test )
+RESOURCES := $(shell  find . | grep -v test )
 
 $(DST): $(RESOURCES) 
 	echo "Building $(NAME)"; pip3 install -r requirements.txt -t . && zip -r -q $(DST) .
```
Comment (rstrahan) [inline lambda/kendra-webcrawler-schedule-updater/Makefile:6]: The Makefiles should, in theory, also install python modules to subdirectory 'py_modules' as we do elsewhere.
It's kind of moot since the requiremenst.txt is empty for them all, so we can defer this to later if you want.

File: lambda/lexv2-build/Makefile [hunk 2]
```diff
@@ -1,6 +1,6 @@
 NAME=$(shell basename $(shell pwd))
 DST=../../build/lambda/$(NAME).zip
-
+RESOURCES := $(shell  find . | grep -v test )
 
 $(DST): $(RESOURCES) 
-	echo "Building $(NAME)"; pip3 install -r requirements.txt -t . && zip -r -q $(DST) .
+	echo "Building $(NAME)"; pip3 install -r requirements.txt -t ./py_modules && zip -r -q $(DST) .
```
Comment (rstrahan) [inline lambda/lexv2-build/Makefile:6]: This also seems unrelated to PR title - "Kendra webcrawler fixes"  - probably should have been in a separate PR for clarity, but doesn't really matter.

**Review by rstrahan — APPROVED** [review]
Looks good Raj.. See comments.. no blockers though.. Thanks!

File: lambda/lexv2-build/Makefile (hunk 2 above)
Comment (rajc-aws) [reply lambda/lexv2-build/Makefile:6]: sorry -- yes. i should have created a separate PR for this..

File: lambda/kendra-webcrawler-schedule-updater/Makefile (hunk 1 above)
Comment (rajc-aws) [reply lambda/kendra-webcrawler-schedule-updater/Makefile:6]: good point. I think we should take a look at all our makefiles and update. I did a scan for a few, but i think we should review this more.

File: lambda/kendra-webcrawler-status/kendra_webcrawler_status.py [hunk 3]
```diff
@@ -48,31 +48,41 @@ def get_data_source_id(index_id, data_source_name):
...
+    if response['History'] != []:   #if there is data sync history
+        latest_history_item = functools.reduce(lambda x, y: x if x['StartTime'] > y['StartTime'] else y, response['History'])
+
+        result = list(map(lambda item: {'StartTime': item['StartTime'].strftime("%m/%d/%Y, %H:%M:%S"),
                                     'EndTime': item['EndTime'].strftime("%m/%d/%Y, %H:%M:%S") if 'EndTime' in item else '',
                                     'Status': item['Status'] if item['Status'] != "INCOMPLETE" else "COMPLETE WITH ERRORS",
                                     'ErrorMessage': item['ErrorMessage'] if 'ErrorMessage' in item else '',
                                     'Metrics': item['Metrics']
                                     }, response['History']))
+        status = latest_history_item['Status']
```
Comment (aassadza) [inline lambda/kendra-webcrawler-status/kendra_webcrawler_status.py:73]: Are we overloading the `status` of the data source,

-  `'Status': 'CREATING'|'DELETING'|'FAILED'|'UPDATING'|'ACTIVE',`

 and the `status` of data source sync job (from the latest history of the sync job)

-  `'Status': 'FAILED'|'SUCCEEDED'|'SYNCING'|'INCOMPLETE'|'STOPPING'|'ABORTED'|'SYNCING_INDEXING',`

And the superset of values is returned in the response of this function?

**Review by aassadza — APPROVED** [review]

File: lambda/kendra-webcrawler-schedule-updater/kendra_webcrawler_schedule_updater.py [hunk 4]
```diff
@@ -130,10 +131,13 @@ def kendra_update_data_source(IndexId, data_source_id, URLs, RoleArn, schedule,
                 'Urls': {
                     'SeedUrlConfiguration': {
                         'SeedUrls': URLs,
-                        'WebCrawlerMode': 'EVERYTHING'
+                        'WebCrawlerMode': crawler_mode
                     }
                 },
-                'CrawlDepth': int(crawl_depth)
+                'CrawlDepth': int(crawl_depth),
```
[the remark this answers is not in the document]
Comment (rajc-aws) [reply lambda/kendra-webcrawler-schedule-updater/kendra_webcrawler_schedule_updater.py:137]: Good call out. Currently using default values, but we should add these additional parameters as a customizable option in the UI to simplify setups. And also display these for reference when on the Web Crawler page.

File: lambda/kendra-webcrawler-status/kendra_webcrawler_status.py [hunk 5]
```diff
@@ -48,31 +48,41 @@ def get_data_source_id(index_id, data_source_name):
...
+    #get information about a Kendra index data source
+    response = client.describe_data_source(
         Id=data_source_id,
-        IndexId=IndexId,
+        IndexId=IndexId
     )
+    data_source_status = response["Status"]
 
+    #get information on the data sync jobs for a given data source
+    response = client.list_data_source_sync_jobs(
```
[the remark this answers is not in the document]
Comment (rajc-aws) [reply lambda/kendra-webcrawler-status/kendra_webcrawler_status.py:59]: Yes, we should review this more and see how best to display sync jobs (current run, vs. previous runs).

File: lambda/kendra-webcrawler-status/kendra_webcrawler_status.py (hunk 3 above)
Comment (rajc-aws) [reply lambda/kendra-webcrawler-status/kendra_webcrawler_status.py:73]: We are using two different status(es) for this -- Data Source status, and Data Source Sync Job status.
But yes, there is room here to make this more clearer in the UI, when displaying status(es). Lets review this more.
